proposals/custom-states-and-state-pseudo-class.md

```diff
@@ -0,0 +1,105 @@
+
+### Goals
+
+-   Give a way for custom element authors to expose varying states of the custom elements to custom element users
+-   Allow styling of custom elements that differ according state
+-   Allow styling of custom elements that differ according to state even when they are only accessible via shadow parts. That is, `element::part(x):state(y)` should work if `::part(x)` selects a custom element with state y..
+    
+
+### Non-goals
+
+-   Allowing adding custom states to built-in elements. States reflect outward what the element implementation author wishes to show the world; for application-level state, use the `class=""` attribute or similar.
+-   Allowing custom element authors to modify already-existing/built-in states/pseudo classes, e.g. `:checked`. (This might be added in the future, though.)
+
+## Example
+
+```html	
+<!DOCTYPE html>
+<body>
+<script>
+class LabeledCheckbox extends  HTMLElement {
+  constructor() {
+    super();
+    this._internals = this.attachInternals();
+    this._checked = false;
+    this.addEventListener('click', this._onClick.bind(this));
+
+    this._shadowRoot = this.attachShadow({mode: "open"});
+    this._shadowRoot.innerHTML =
+      "<style>\
+        :host::before { content: '[]'; }\
+        :host(:state(checked))::before { content: '[x]' }\
+      </style>\
+      <span><slot>Label</slot></span>";
+  }
+  
+  get checked() { return this._checked; }
+  
+  set checked(flag) {
+    this._checked = !!flag;
+    if (this._checked) {
+      this._internals.states.add("checked");
+    } else {
+      this._internals.states.remove("checked");
+    }
+  }
+
+  _onClick(event) {
+    this.checked = !this._checked;
+  }
+  
+}
+  
+class QuestionBox extends HTMLElement {
+  constructor() {
+    super();
+    this._shadowRoot = this.attachShadow({mode: "open"});
+    this._shadowRoot.innerHTML =
+      "<div><slot>Do you agree?</slot></div>\
+      <labeled-checkbox part='checkbox'>Yes</labeled-checkbox>";
+  }
+}
+customElements.define('labeled-checkbox', LabeledCheckbox);
+customElements.define('question-box', QuestionBox);
+</script>
+
+<style>
+  question-box::part(checkbox){ color: red;}
+  question-box::part(checkbox):state(checked) { color: green; }
+</style>
+  
+<question-box>Continue?</question-box>
+</body>
+```
+## Proposal 
+
+Add a `states` property to the [ElementInternals](https://html.spec.whatwg.org/multipage/custom-elements.html#elementinternals) interface to contain a list of states for the corresponding custom element, and a new `:state(x)` pseudo-class that can select custom elements that contains `x` in its `ElementInternals`' `state`. An example implementation of a custom element that uses this is shown above.
```

Comment:
The code example indicates `states` has `add()` and `remove()`.  What other operations does it have?  Is it a `DOMTokenList`?